File: keevo/app/lib/features/pos/domain/usecase/get_sales_history_usecase.dart
import '../model/sale_model.dart';
import '../model/sales_history_filter.dart';
import '../repository/sale_repository.dart';

/// GetSalesHistoryUseCase — Retrieves sales history with filters.
///
/// Story 4.4 — Clôture Journalière & Historique des Ventes.
class GetSalesHistoryUseCase {
  final SaleRepository _repository;

  GetSalesHistoryUseCase({required SaleRepository repository})
      : _repository = repository;

  /// Execute the sales history query.
  ///
  /// Returns sales matching the filter criteria, sorted by occurredAt DESC.
  /// For EMPLOYEE role, only returns own sales.
  /// For OWNER role, returns all store sales (can optionally filter by employee).
  Future<List<Sale>> execute(SalesHistoryFilter filter) async {
    return _repository.getSalesHistory(filter);
  }
}

File: keevo/backend/src/main/java/com/keevo/commerce/sale/domain/port/in/GetSalesHistoryUseCase.java
package com.keevo.commerce.sale.domain.port.in;

import com.keevo.commerce.sale.domain.model.Sale;
import org.springframework.data.domain.Page;

import java.time.Instant;
import java.util.UUID;

/**
 * GetSalesHistoryUseCase — Port for retrieving sales history with filters.
 * Story 4.4 — Clôture Journalière & Historique des Ventes
 *
 * <p>EMPLOYEE: sees only their own sales (employeeId forced to JWT actorId).
 * <p>OWNER: sees all sales in the store (employeeId filter is optional).
 */
public interface GetSalesHistoryUseCase {

    /**
     * Query for sales history.
     *
     * @param storeId    Store to query
     * @param employeeId Optional employee filter (ignored for EMPLOYEE role — forced to JWT actorId)
     * @param from       Start of date range (inclusive)
     * @param to         End of date range (inclusive)
     * @param role       User role: "OWNER" or "EMPLOYEE"
     * @param page       Page number (0-indexed)
     * @param size       Page size
     */
    record SalesHistoryQuery(UUID storeId, UUID employeeId, Instant from, Instant to,
                             String role, int page, int size) {}

    /**
     * Get paginated sales history.
     *
     * @param query search criteria
     * @return page of sales matching the criteria
     */
    Page<Sale> getSalesHistory(SalesHistoryQuery query);
}
